pkg/core/oracle/oracle.go

```diff
@@ -0,0 +1,286 @@
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/config/netmode"
+	"github.com/nspcc-dev/neo-go/pkg/core/blockchainer"
+	"github.com/nspcc-dev/neo-go/pkg/core/native"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/wallet"
+	"go.uber.org/zap"
+)
+
+type (
+	// Oracle represents oracle module capable of talking
+	// with the external world.
+	Oracle struct {
+		oracleScript []byte
+		oracleHash   util.Uint160
+
+		accMtx      sync.RWMutex
+		currAccount *wallet.Account
+		oracleNodes keys.PublicKeys
+
+		responses map[uint64]*incompleteTx
+
+		log     *zap.Logger
+		wallet  *wallet.Wallet
+		client  HTTPClient
+		network netmode.Magic
+
+		bc            blockchainer.Blockchainer
+		onResponse    ResponseCallback
+		onTransaction TxCallback
+		uriValidator  URIValidator
+	}
+
+	// HTTPClient is an interface capable of doing oracle requests.
+	HTTPClient interface {
+		Get(string) (*http.Response, error)
+	}
+
+	// ResponseCallback signs and broadcast this node oracle responses.
+	ResponseCallback = func(priv *keys.PrivateKey, resp *transaction.OracleResponse, txSig []byte)
+	// TxCallback executes on new transactions when they are ready to be pooled.
+	TxCallback = func(tx *transaction.Transaction)
+	// URIValidator is used to check if provided URL is valid.
+	URIValidator = func(*url.URL) error
+)
+
+const (
+	// MaxAllowedResponse is a maximum size of response in bytes.
+	MaxAllowedResponse = math.MaxUint16
+
+	// requestTimeout is default request timeout.
+	requestTimeout = time.Millisecond * 500
+)
+
+// NewOracle returns new oracle instance.
+func NewOracle(oracleScript []byte, cfg config.Config, log *zap.Logger) (*Oracle, error) {
+	o := &Oracle{
+		oracleScript: oracleScript,
+		oracleHash:   hash.Hash160(oracleScript),
+
+		responses: make(map[uint64]*incompleteTx),
+		log:       log,
+		network:   cfg.ProtocolConfiguration.Magic,
+	}
+
+	var err error
+	w := cfg.ApplicationConfiguration.UnlockWallet
+	if o.wallet, err = wallet.NewWalletFromFile(w.Path); err != nil {
+		return nil, err
+	}
+	for _, acc := range o.wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err != nil {
+			return nil, err
+		}
+	}
+
+	var client http.Client
+	client.Transport = &http.Transport{DisableKeepAlives: true}
+	client.Timeout = requestTimeout
+	o.client = &client
+
+	// nop by default.
+	o.SetResponseCallback(func(*keys.PrivateKey, *transaction.OracleResponse, []byte) {})
+	o.SetTxCallback(func(*transaction.Transaction) {})
+	o.SetURIValidator(defaultURIValidator)
+	return o, nil
+}
+
+// SetHTTPClient sets http client for oracle responses.
+func (o *Oracle) SetHTTPClient(client HTTPClient) {
+	o.client = client
+}
+
+// SetURIValidator sets URI validator.
+func (o *Oracle) SetURIValidator(v URIValidator) {
```

Comment:
Because urls used in tests are mocked and checking if URL is valid requires getting IP from hostname. I will make it a part of config passed to `NewOracle()`.

Reply:
Fixed.